signer/src/utxo.rs

```diff
@@ -0,0 +1,964 @@
+use bitcoin::ScriptBuf;
+use bitcoin::Sequence;
+use bitcoin::Transaction;
+use bitcoin::TxIn;
+use bitcoin::TxOut;
+use bitcoin::Witness;
+use bitcoin::XOnlyPublicKey;
+use secp256k1::Keypair;
+use secp256k1::Message;
+
+use crate::error::Error;
+use crate::packaging::compute_optimal_packages;
+use crate::packaging::Weighted;
+
+#[derive(Debug, Clone, Copy)]
+pub struct SignerBtcState {
+    /// The outstanding signer UTXO.
+    pub utxo: SignerUtxo,
+    /// The current market fee rate in sat/vByte.
+    pub fee_rate: u64,
+    /// The current public key of the signers
+    pub public_key: PublicKey,
+}
+
+#[derive(Debug)]
+pub struct SbtcRequests {
+    /// Accepted and pending deposit requests.
+    pub deposits: Vec<DepositRequest>,
+    /// Accepted and pending withdrawal requests.
+    pub withdrawals: Vec<WithdrawalRequest>,
+    /// Summary of the Signers' UTXO and information necessary for
+    /// constructing their next UTXO.
+    pub signer_state: SignerBtcState,
+    /// The maximum acceptable number of votes against for any given
+    /// request.
+    pub reject_capacity: u32,
+}
+
+impl SbtcRequests {
+    /// Construct the next transaction package given requests and the
+    /// signers' UTXO.
+    ///
+    /// This function can fail if the output amounts are greater than the
+    /// input amounts.
+    pub fn construct_transactions(&self) -> Result<Vec<UnsignedTransaction>, Error> {
+        if self.deposits.is_empty() && self.withdrawals.is_empty() {
+            tracing::info!("No deposits or withdrawals so no BTC transaction");
+            return Ok(Vec::new());
+        }
+
+        let withdrawals = self.withdrawals.iter().map(Request::Withdrawal);
+        let deposits = self.deposits.iter().map(Request::Deposit);
+
+        // Create a list of requests where each request can be approved on its own.
+        let items = deposits.chain(withdrawals);
+
+        compute_optimal_packages(items, self.reject_capacity)
+            .scan(self.signer_state, |state, requests| {
+                let tx = UnsignedTransaction::new(requests, state);
+                if let Ok(tx_ref) = tx.as_ref() {
+                    state.utxo = tx_ref.new_signer_utxo();
+                }
+                Some(tx)
+            })
+            .collect()
+    }
+}
+
+#[derive(Debug)]
+pub struct DepositRequest {
+    /// The UTXO to be spent by the signers.
+    pub outpoint: OutPoint,
+    /// The max fee amount to use for the BTC deposit transaction.
+    pub max_fee: u64,
+    /// How each of the signers voted for the transaction.
+    pub signer_bitmap: Vec<bool>,
+    /// The amount of sats in the deposit UTXO.
+    pub amount: u64,
+    /// The deposit script used so that the signers' can spend funds.
+    pub deposit_script: ScriptBuf,
+    /// The redeem script for the deposit.
+    pub redeem_script: ScriptBuf,
+    /// The public key used for the key-spend path of the taproot script.
+    pub taproot_public_key: PublicKey,
```

Comment:
A deposit request doesn't have a key-spend path (as specified in the [design issue](https://github.com/stacks-network/sbtc/issues/30)). We don't want to allow anyone to spend the deposit request while the signers are processing it. Instead, the key-spend should be the NUMS point from [BIP-341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki) $H = lift_x(0x50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0)$

Therefore, I don't think this should be tracked in the `DepositRequest` struct, but rather be a hard-coded constant.

Reply:
Right, I see you've already documented this below that it should be a dummy point. I assume you then expect the validation of this point to happen before it's used in the `construct_witness_data` function then?

Reply:
> Right, I see you've already documented this below that it should be a dummy point. I assume you then expect the validation of this point to happen before it's used in the `construct_witness_data` function then?

Sorta. I've been assuming both the signers and the Deposit API would do this validation. But it would all happen using the Transaction Analysis library when constructing the `DepositRequest` struct here.

Reply:
> Therefore, I don't think this should be tracked in the `DepositRequest` struct, but rather be a hard-coded constant.

Okay, I'll hard code it in the `construct_witness_data` function.

Reply:
So after a quick sync we've agreed to do this as a follow-up instead because:

1. It's not trivial
2. It's not obvious if the validation of the NUMS point should be the concern of this particular module.

Reply:
I created https://github.com/stacks-network/sbtc/issues/130 for this work.

Reply:
Perfect, thanks!